reactor-netty-http/src/main/java/reactor/netty/http/client/Http2ConnectionLivenessHandler.java

```diff
@@ -0,0 +1,285 @@
+		}
+		else {
+			this.pingScheduleIntervalNanos = 0L;
+		}
+
+		if (pingAckDropThreshold != null) {
+			this.pingAckDropThreshold = pingAckDropThreshold;
+		}
+		else {
+			this.pingAckDropThreshold = 0;
+		}
+	}
+
+	@Override
+	public void channelActive(ChannelHandlerContext ctx) throws Exception {
+		if (isPingIntervalConfigured()) {
+			isPingAckPending = false;
+			pingAckDropCount = 0;
+			pingScheduler = ctx.executor()
+					.schedule(
+							new PingChecker(ctx),
+							pingAckTimeoutNanos,
+							NANOSECONDS
+					);
+		}
+
+		ctx.fireChannelActive();
+	}
+
+	@Override
+	public void channelRead(ChannelHandlerContext ctx, Object msg) throws Exception {
+		if (msg instanceof Http2PingFrame) {
+			Http2PingFrame frame = (Http2PingFrame) msg;
+			if (frame.ack() && frame.content() == lastSentPingData) {
+				lastReceivedPingTime = System.nanoTime();
+			}
+		}
+		else {
+			lastIoTime = System.nanoTime();
+		}
+
+		ctx.fireChannelRead(msg);
+	}
+
+	@Override
+	@SuppressWarnings("FutureReturnValueIgnored")
+	public void write(ChannelHandlerContext ctx, Object msg, ChannelPromise promise) throws Exception {
+		lastIoTime = System.nanoTime();
+
+		ctx.write(msg, promise);
+	}
+
+	@Override
+	public void channelInactive(ChannelHandlerContext ctx) throws Exception {
+		cancel();
+		ctx.fireChannelInactive();
+	}
+
+	private boolean isPingIntervalConfigured() {
+		return pingAckTimeoutNanos > 0
+				&& pingScheduleIntervalNanos > 0;
+	}
+
+	private void cancel() {
+		if (pingScheduler != null) {
+			pingScheduler.cancel(false);
+		}
+	}
+
+	private class PingChecker implements Runnable {
+
+		private final ChannelHandlerContext ctx;
+
+		PingChecker(ChannelHandlerContext ctx) {
+			this.ctx = ctx;
+		}
+
+		@Override
+		public void run() {
+			Channel channel = ctx.channel();
+			if (channel == null || !channel.isOpen()) {
+				return;
+			}
+
+			if (lastIoTime == 0 || isIoInProgress()) {
+				if (log.isDebugEnabled()) {
+					log.debug("{} channel is currently reading or writing data.", channel);
+				}
+
+				isPingAckPending = false;
+				pingAckDropCount = 0;
+				pingScheduler = invokeNextSchedule();
+				return;
+			}
+
+			if (!isPingAckPending) {
+				if (log.isDebugEnabled()) {
+					log.debug("Attempting to send a ping frame to {} channel.", channel);
+				}
+
+				writePing(ctx);
+				pingScheduler = invokeNextSchedule();
+				return;
+			}
+
+			if (isOutOfTimeRange()) {
+				countPingDrop();
+
+				if (isExceedAckDropThreshold()) {
+					if (log.isInfoEnabled()) {
+						log.info("Closing {} channel due to delayed ping frame response (timeout: {} ns). lastReceivedPingTime: {}, current: {}", channel, pingAckTimeoutNanos, lastReceivedPingTime, System.nanoTime());
+					}
+
+					close(channel);
+					return;
+				}
+
+				if (log.isInfoEnabled()) {
+					log.info("Drop ping ack frame in {} channel. (ping: {})", channel, lastSentPingData);
+				}
+
+				writePing(ctx);
+				pingScheduler = invokeNextSchedule();
+				return;
+			}
```

Comment:
If the **Ping ACK is not received** within `pingAckTimeoutNanos`, but the retry count has not yet reached the `pingAckDropThreshold`, a **retry is attempted**.